File: Flutter/Flutter/android/app/src/main/java/com/example/flutter_gertec/GertecPrinter.java
package com.example.flutter_gertec;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Paint;
import android.graphics.Typeface;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;
import com.journeyapps.barcodescanner.BarcodeEncoder;

import br.com.gertec.gedi.GEDI;
import br.com.gertec.gedi.enums.GEDI_PRNTR_e_Alignment;
import br.com.gertec.gedi.enums.GEDI_PRNTR_e_BarCodeType;
import br.com.gertec.gedi.enums.GEDI_PRNTR_e_Status;
import br.com.gertec.gedi.exceptions.GediException;
import br.com.gertec.gedi.interfaces.ICL;
import br.com.gertec.gedi.interfaces.IGEDI;
import br.com.gertec.gedi.interfaces.IPRNTR;
import br.com.gertec.gedi.structs.GEDI_PRNTR_st_BarCodeConfig;
import br.com.gertec.gedi.structs.GEDI_PRNTR_st_PictureConfig;
import br.com.gertec.gedi.structs.GEDI_PRNTR_st_StringConfig;

public class GertecPrinter {

    ICL icl = null;
    // Definições
    private final String IMPRESSORA_ERRO = "Impressora com erro.";

    // Statics
    private static boolean isPrintInit = false;

    // Vaviáveis iniciais
    private Activity activity;
    private Context context;

    // Classe de impressão
    private IGEDI iGedi = null;
    private IPRNTR iPrint = null;
    private GEDI_PRNTR_st_StringConfig stringConfig;
    private GEDI_PRNTR_st_PictureConfig pictureConfig;
    private GEDI_PRNTR_e_Status status;

    // Classe de configuração da impressão
    private ConfigPrint configPrint;
    private Typeface typeface;

    /**
     * Método construtor da classe
     * 
     * @param c = Context atual que esta sendo inicializada a class
     */
    public GertecPrinter(Context c) {
        this.context = c;
        startIGEDI();
    }

    /**
     * Método construtor da classe
     * 
     * @param a = Activity atual que esta sendo inicializada a class
     *
     *          public GertecPrinter(Activity a) { this.activity = a; startIGEDI(a);
     *          }
     */
    // public GertecPrinter(Activity a) {
    // this.activity = a;
    // startIGEDI(a);
    // }

    /**
     * Método que instância a classe GEDI da lib
     *
     * @apiNote = Este mátodo faz a instância da classe GEDI através de uma Thread.
     *          Será sempre chamado na construção da classe. Não alterar...
     */
    private void startIGEDI() {
        new Thread(() -> {
            GEDI.init(this.context);
            this.iGedi = GEDI.getInstance(this.context);
            this.iPrint = this.iGedi.getPRNTR();
            icl = GEDI.getInstance().getCL(); // Get ICL
            try {
                new Thread().sleep(250);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }).start();
    }

    /**
     * Método que instância a classe GEDI da lib
     *
     * @apiNote = Este mátodo faz a instância da classe GEDI através de uma Thread.
     *          Será sempre chamado na construção da classe. Não alterar...
     * 
     *          private void startIGEDI(Activity a) { new Thread(() -> { iGedi = new
     *          Gedi(a); this.iGedi = GEDI.getInstance(a); this.iPrint =
     *          this.iGedi.getPRNTR(); try { new Thread().sleep(250); } catch
     *          (InterruptedException e) { e.printStackTrace(); } }).start(); }
     */
    // private void startIGEDI(Activity a) {
    // new Thread(() -> {
    // iGedi = new Gedi(a);
    // this.iGedi = GEDI.getInstance(a);
    // this.iPrint = this.iGedi.getPRNTR();
    // try {
    // new Thread().sleep(250);
    // } catch (InterruptedException e) {
    // e.printStackTrace();
    // }
    // }).start();
    // }
    /**
     * Método que recebe a configuração para ser usada na impressão
     * 
     * @param config = Classe {@link ConfigPrint} que contém toda a configuração
     *               para a impressão
     */
    public void setConfigImpressao(ConfigPrint config) {

        this.configPrint = config;

        this.stringConfig = new GEDI_PRNTR_st_StringConfig(new Paint());
        this.stringConfig.paint.setTextSize(configPrint.getTamanho());
        this.stringConfig.paint.setTextAlign(Paint.Align.valueOf(configPrint.getAlinhamento()));
        this.stringConfig.offset = configPrint.getOffSet();
        this.stringConfig.lineSpace = configPrint.getLineSpace();

        switch (configPrint.getFonte()) {
            case "NORMAL":
                this.typeface = Typeface.create(configPrint.getFonte(), Typeface.NORMAL);
                break;
            case "DEFAULT":
                this.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL);
                break;
            case "DEFAULT BOLD":
                this.typeface = Typeface.create(Typeface.DEFAULT_BOLD, Typeface.NORMAL);
                break;
            case "MONOSPACE":
                this.typeface = Typeface.create(Typeface.MONOSPACE, Typeface.NORMAL);
                break;
            case "SANS SERIF":
                this.typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL);
                break;
            case "SERIF":
                this.typeface = Typeface.create(Typeface.SERIF, Typeface.NORMAL);
                break;
            default:
                this.typeface = Typeface.createFromAsset(this.context.getAssets(), configPrint.getFonte());
        }

        if (this.configPrint.isNegrito() && this.configPrint.isItalico()) {
            typeface = Typeface.create(typeface, Typeface.BOLD_ITALIC);
        } else if (this.configPrint.isNegrito()) {
            typeface = Typeface.create(typeface, Typeface.BOLD);
        } else if (this.configPrint.isItalico()) {
            typeface = Typeface.create(typeface, Typeface.ITALIC);
        }

        if (this.configPrint.isSublinhado()) {
            this.stringConfig.paint.setFlags(Paint.UNDERLINE_TEXT_FLAG);
        }

        this.stringConfig.paint.setTypeface(this.typeface);
    }

    /**
     * Método que retorna o atual estado da impressora
     *
     * @throws GediException = vai retorno o código do erro.
     *
     * @return String = traduzStatusImpressora()
     *
     */
    public String getStatusImpressora() throws GediException {
        try {
            ImpressoraInit();
            this.status = this.iPrint.Status();
        } catch (GediException e) {
            throw new GediException(e.getErrorCode());
        }

        return traduzStatusImpressora(this.status);
    }

    /**
     * Método que recebe o atual texto a ser impresso
     * 
     * @param texto = Texto que será impresso.
     *
     * @throws Exception = caso a impressora esteja com erro.
     *
     */
    public void imprimeTexto(String texto) throws Exception {

        // this.getStatusImpressora();
        try {
            if (!isImpressoraOK()) {
                throw new Exception(IMPRESSORA_ERRO);
            }
            sPrintLine(texto);
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * Método que recebe o atual texto e o tamanho da fonte que deve ser usado na
     * impressão.
     *
     * @param texto   = Texto que será impresso.
     * @param tamanho = Tamanho da fonte que será usada
     *
     * @throws Exception = caso a impressora esteja com erro.
     *
     * @apiNote = Esse mátodo só altera o tamanho do texto na impressão que for
     *          chamado a classe {@link ConfigPrint} não será alterada para
     *          continuar sendo usado na impressão da proxíma linha
     *
     */
    public void imprimeTexto(String texto, int tamanho) throws Exception {

        int tamanhoOld;

        // this.getStatusImpressora();

        try {
            if (!isImpressoraOK()) {
                throw new Exception(IMPRESSORA_ERRO);
            }
            tamanhoOld = this.configPrint.getTamanho();
            this.configPrint.setTamanho(tamanho);
            sPrintLine(texto);
            this.configPrint.setTamanho(tamanhoOld);
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * Método que recebe o atual texto e ser o mesmo será impresso em negrito.
     *
     * @param texto   = Texto que será impresso.
     * @param negrito = Caso o texto deva ser impresso em negrito
     *
     * @throws Exception = caso a impressora esteja com erro.
     *
     * @apiNote = Esse mátodo só altera o tamanho do texto na impressão que for
     *          chamado * a classe {@link ConfigPrint} não será alterada para
     *          continuar sendo usado na impressão da * proxíma linha
     *
     */
    public void imprimeTexto(String texto, boolean negrito) throws Exception {

        boolean negritoOld = false;

        // this.getStatusImpressora();

        try {
            if (!isImpressoraOK()) {
                throw new Exception(IMPRESSORA_ERRO);
            }
            negritoOld = this.configPrint.isNegrito();
            this.configPrint.setNegrito(negrito);

            sPrintLine(texto);

            this.configPrint.setNegrito(negritoOld);

        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * Método que recebe o atual texto e ser o mesmo será impresso em negrito e/ou
     * itálico.
     *
     * @param texto   = Texto que será impresso.
     * @param negrito = Caso o texto deva ser impresso em negrito
     * @param italico = Caso o texto deva ser impresso em itálico
     *
     * @throws Exception = caso a impressora esteja com erro.
     *
     * @apiNote = Esse mátodo só altera o tamanho do texto na impressão que for
     *          chamado * a classe {@link ConfigPrint} não será alterada para
     *          continuar sendo usado na impressão da * proxíma linha
     *
     */
    public void imprimeTexto(String texto, boolean negrito, boolean italico) throws Exception {

        boolean negritoOld = false;
        boolean italicoOld = false;

        // this.getStatusImpressora();

        try {
            if (!isImpressoraOK()) {
                throw new Exception(IMPRESSORA_ERRO);
            }
            negritoOld = this.configPrint.isNegrito();
            italicoOld = this.configPrint.isItalico();
            this.configPrint.setNegrito(negrito);
            this.configPrint.setItalico(italico);

            sPrintLine(texto);

            this.configPrint.setNegrito(negritoOld);
            this.configPrint.setItalico(italicoOld);

        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * Método que recebe o atual texto e ser o mesmo será impresso em negrito,
     * itálico e/ou sublinhado.
     *
     * @param texto      = Texto que será impresso.
     * @param negrito    = Caso o texto deva ser impresso em negrito
     * @param italico    = Caso o texto deva ser impresso em itálico
     * @param sublinhado = Caso o texto deva ser impresso em itálico.
     *
     * @throws Exception = caso a impressora esteja com erro.
     *
     * @apiNote = Esse mátodo só altera o tamanho do texto na impressão que for
     *          chamado * a classe {@link ConfigPrint} não será alterada para
     *          continuar sendo usado na impressão da * proxíma linha
     *
     */
    public void imprimeTexto(String texto, boolean negrito, boolean italico, boolean sublinhado) throws Exception {

        boolean negritoOld = false;
        boolean italicoOld = false;
        boolean sublinhadoOld = false;

        // this.getStatusImpressora();

        try {
            if (!isImpressoraOK()) {
                throw new Exception(IMPRESSORA_ERRO);
            }
            negritoOld = this.configPrint.isNegrito();
            italicoOld = this.configPrint.isItalico();
            sublinhadoOld = this.configPrint.isSublinhado();

            this.configPrint.setNegrito(negrito);
            this.configPrint.setItalico(italico);
            this.configPrint.setSublinhado(sublinhado);

            sPrintLine(texto);

            this.configPrint.setNegrito(negritoOld);
            this.configPrint.setItalico(italicoOld);
            this.configPrint.setSublinhado(sublinhadoOld);

        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * Método privado que faz a impressão do texto.
     *
     * @param texto = Texto que será impresso
     *
     * @throws GediException = retorna o código do erro
     *
     */
    private boolean sPrintLine(String texto) throws Exception {
        // Print Data
        try {
            ImpressoraInit();
            this.iPrint.DrawStringExt(this.stringConfig, texto);
            this.avancaLinha(configPrint.getAvancaLinhas());
            // ImpressoraOutput();
            return true;
        } catch (GediException e) {
            throw new GediException(e.getErrorCode());
        }
    }

    /**
     * Método que faz a impressão de imagens
     *
     * @param imagem = Nome da imagem que deve estar na pasta drawable
     *
     * @throws IllegalArgumentException = Argumento passado ilegal
     * @throws GediException            = retorna o código do erro.
     *
     */
    public boolean imprimeImagem(String imagem) throws GediException {

        int id = 0;
        Bitmap bmp;
        try {

            pictureConfig = new GEDI_PRNTR_st_PictureConfig();

            // Align
            pictureConfig.alignment = GEDI_PRNTR_e_Alignment.valueOf(configPrint.getAlinhamento());

            // Height
            pictureConfig.height = this.configPrint.getiHeight();
            // Width
            pictureConfig.width = this.configPrint.getiWidth();

            if (MainActivity.Model.equals(MainActivity.G700)) {
                id = context.getResources().getIdentifier(imagem, "drawable", context.getPackageName());
                bmp = BitmapFactory.decodeResource(context.getResources(), id);
            } else {
                id = this.activity.getApplicationContext().getResources().getIdentifier(imagem, "drawable",
                        this.activity.getApplicationContext().getPackageName());
                bmp = BitmapFactory.decodeResource(this.activity.getResources(), id);
            }

            ImpressoraInit();
            this.iPrint.DrawPictureExt(pictureConfig, bmp);
            this.avancaLinha(configPrint.getAvancaLinhas());

            return true;
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(e);
        } catch (GediException e) {
            throw new GediException(e.getErrorCode());
        }

    }

    /**
     * Método que faz a impressão de código de barras
     *
     * @param texto       = Texto que será usado para a impressão do código de
     *                    barras
     * @param height      = Tamanho
     * @param width       = Tamanho
     * @param barCodeType = Tipo do código que será impresso
     *
     * @throws IllegalArgumentException = Argumento passado ilegal
     * @throws GediException            = retorna o código do erro.
     *
     */
    public boolean imprimeBarCode(String texto, int height, int width, String barCodeType) throws GediException {

        try {

            GEDI_PRNTR_st_BarCodeConfig barCodeConfig = new GEDI_PRNTR_st_BarCodeConfig();
            // Bar Code Type
            barCodeConfig.barCodeType = GEDI_PRNTR_e_BarCodeType.valueOf(barCodeType);

            // Height
            barCodeConfig.height = height;
            // Width
            barCodeConfig.width = width;

            ImpressoraInit();
            this.iPrint.DrawBarCode(barCodeConfig, texto);
            this.avancaLinha(configPrint.getAvancaLinhas());
            // ImpressoraOutput();
            // this.iPrint.Output();
            return true;
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(e);
        } catch (GediException e) {
            throw new GediException(e.getErrorCode());
        }

    }

    /**
     * Método que faz a impressão de código de barras
     *
     * @param texto       = Texto que será usado para a impressão do código de
     *                    barras
     * @param height      = Tamanho
     * @param width       = Tamanho
     * @param barCodeType = Tipo do código que será impresso
     *
     * @throws IllegalArgumentException = Argumento passado ilegal
     * @throws GediException            = retorna o código do erro.
     *
     */
    public boolean imprimeBarCodeIMG(String texto, int height, int width, String barCodeType)
            throws GediException, WriterException {

        try {

            MultiFormatWriter multiFormatWriter = new MultiFormatWriter();
            BitMatrix bitMatrix = multiFormatWriter.encode(texto, BarcodeFormat.valueOf(barCodeType), height, width);
            BarcodeEncoder barcodeEncoder = new BarcodeEncoder();
            Bitmap bitmap = barcodeEncoder.createBitmap(bitMatrix);

            pictureConfig = new GEDI_PRNTR_st_PictureConfig();
            pictureConfig.alignment = GEDI_PRNTR_e_Alignment.valueOf(configPrint.getAlinhamento());

            pictureConfig.height = bitmap.getHeight();
            pictureConfig.width = bitmap.getWidth();

            ImpressoraInit();
            this.iPrint.DrawPictureExt(pictureConfig, bitmap);

            return true;
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
            throw new IllegalArgumentException(e);
        } catch (GediException e) {
            e.printStackTrace();
            throw new GediException(e.getErrorCode());
        } catch (WriterException e) {
            e.printStackTrace();
            throw new WriterException(e);
        }

    }

    /**
     * Método que faz o avanço de linhas após uma impressão.
     *
     * @param linhas = Número de linhas que dever ser pulado após a impressão.
     *
     * @throws GediException = retorna o código do erro.
     *
     * @apiNote = Esse método não deve ser chamado dentro de um FOR ou WHILE, o
     *          número de linhas deve ser sempre passado no atributo do método.
     *
     */
    public void avancaLinha(int linhas) throws GediException {
        try {
            if (linhas > 0) {
                this.iPrint.DrawBlankLine(linhas);
            }
        } catch (GediException e) {
            throw new GediException(e.getErrorCode());
        }
    }

    /**
     * Método que retorno se a impressora está apta a fazer impressões
     *
     * @return true = quando estiver tudo ok.
     *
     */
    public boolean isImpressoraOK() {

        if (status.getValue() == 0) {
            return true;
        }
        return false;
    }

    /**
     * Método que faz a inicialização da impressao
     *
     * @throws GediException = retorno o código do erro.
     *
     */
    public void ImpressoraInit() throws GediException {
        try {
            if (this.iPrint != null && !isPrintInit) {
                this.icl.PowerOff(); // Desligar Módulo NFC - comando Mandatório antes de enviar comandos para a impressora."
                this.iPrint.Init();
                isPrintInit = true;
            }
        } catch (GediException e) {
            e.printStackTrace();
            throw new GediException(e.getErrorCode());
        }
    }

    /**
     * Método que faz a finalizacao do objeto iPrint
     *
     * @throws GediException = retorno o código do erro.
     *
     */
    public void ImpressoraOutput() throws GediException {
        try {
            if (this.iPrint != null) {
                this.iPrint.Output();
                isPrintInit = false;
            }
        } catch (GediException e) {
            e.printStackTrace();
            throw new GediException(e.getErrorCode());
        }
    }

    /**
     * Método que faz a tradução do status atual da impressora.
     *
     * @param status = Recebe o {@link GEDI_PRNTR_e_Status} como atributo
     *
     * @return String = Retorno o atual status da impressora
     *
     */
    private String traduzStatusImpressora(GEDI_PRNTR_e_Status status) {
        String retorno;
        switch (status) {
            case OK:
                retorno = "IMPRESSORA OK";
                break;

            case OUT_OF_PAPER:
                retorno = "SEM PAPEL";
                break;

            case OVERHEAT:
                retorno = "SUPER AQUECIMENTO";
                break;

            default:
                retorno = "ERRO DESCONHECIDO";
                break;
        }

        return retorno;
    }

}

File: Xamarin_Forms/Xamarin_Forms/GertecXamarinForms/GertecXamarinForms.Android/Impressao/GertecPrinter.cs
using Android.App;
using Android.Content;
using Android.Graphics;
using BR.Com.Gertec.Gedi;
using BR.Com.Gertec.Gedi.Enums;
using BR.Com.Gertec.Gedi.Exceptions;
using BR.Com.Gertec.Gedi.Interfaces;
using BR.Com.Gertec.Gedi.Structs;
using Plugin.DeviceInfo;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using ZXing;
using ZXing.Mobile;
using ZXing.Common;

using Xamarin.Forms;


namespace GertecXamarinForms.Droid.Impressao
{
    public class GertecPrinter : IGertecPrinter
    {
        // Defines
        private const string IMPRESSORA_ERRO = "Impressora com erro.";
        
        private string modelo = CrossDeviceInfo.Current.Model;

        // Statics
        private static bool isPrintInit = false;

        private IGEDI iGedi;
        private IPRNTR iPrintr;
        private GEDI_PRNTR_st_StringConfig stringConfig;
        private GEDI_PRNTR_st_PictureConfig pictureConfig;
        private GEDI_PRNTR_e_Status status;

        private ICL icl = null;

        private ConfigPrint configPrint;
        private Typeface typeface;

        // Thread starGedi;
        
        /*private Activity mainActivity;*/
        private Context mainContext;

        public GertecPrinter(Context ctx)
        {
            this.mainContext = ctx;
            startGediGPOS700();
        }

        //  TSG 800
        // public GertecPrinter(Activity act)
        // {
        //     this.mainActivity = act;
        //     startGediTSG800();
        // }

        public void startGediGPOS700()
        {
            new Thread(new ThreadStart(() =>
            {
                GEDI.Init(this.mainContext);
                this.iGedi = GEDI.GetInstance(this.mainContext);
                icl = iGedi.CL;
                this.iPrintr = iGedi.PRNTR;
                Thread.Sleep(100);
            })).Start();
        }

        /*
        public void startGediTSG800()
        {
            new Thread(new ThreadStart(() =>
            {
                #if __G800__
                    this.iGedi = new Gedi(this.mainActivity);
                    this.iGedi = GEDI.GetInstance(this.mainActivity);
                    this.iPrintr = iGedi.PRNTR;
                    Thread.Sleep(250);
                #endif
            })).Start();
        }
        */

        public void ImpressoraInit()
        {
            try
            {
                if (this.iPrintr != null && !isPrintInit)
                {
                    icl.PowerOff(); // Desligar Módulo NFC - comando Mandatório antes de enviar comandos para a impressora.
                    this.iPrintr.Init();
                    isPrintInit = true;
                    getStatusImpressora();
                }
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }
        }

        public void ImpressoraOutput()
        {
            try
            {
                if (this.iPrintr != null && isPrintInit)
                {
                    this.iPrintr.Output();
                    isPrintInit = false;
                }
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }
        }

        public void AvancaLinha(int linhas)
        {
            try
            {
                this.iPrintr.DrawBlankLine(linhas);
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }
        }

        public string getStatusImpressora()
        {
            try
            {
                this.status = this.iPrintr.Status();
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }
            return traduzStatusImpressora(this.status);
        }


        public bool ImprimeBarCode(string texto, int height, int width, string barcodeFormat)
        {
            try
            {

                BarcodeWriter writer = new BarcodeWriter();

                if (barcodeFormat.Equals("CODE_128"))
                {
                    writer.Format = BarcodeFormat.CODE_128;
                }
                else if (barcodeFormat.Equals("EAN_8"))
                {
                    writer.Format = BarcodeFormat.EAN_8;
                }
                else if (barcodeFormat.Equals("EAN_13"))
                {
                    writer.Format = BarcodeFormat.EAN_13;
                }
                else if (barcodeFormat.Equals("PDF_417"))
                {
                    writer.Format = BarcodeFormat.PDF_417;
                }
                else if (barcodeFormat.Equals("QR_CODE"))
                {
                    writer.Format = BarcodeFormat.QR_CODE;
                }

                writer.Options = new EncodingOptions()
                {
                    Width = width,
                    Height = height,
                    Margin = 2
                };

                var bitmap = writer.Write(texto);

                this.pictureConfig = new GEDI_PRNTR_st_PictureConfig();
                this.pictureConfig.Alignment = GEDI_PRNTR_e_Alignment.ValueOf(this.configPrint.Alinhamento);

                this.pictureConfig.Height = bitmap.Height;
                this.pictureConfig.Width = bitmap.Width;

                this.ImpressoraInit();
                this.iPrintr.DrawPictureExt(pictureConfig, bitmap);

                return true;
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }
        }

        public bool ImprimeBarCodeIMG(string texto, int height, int width, string barcodeFormat)
        {
            try
            {
                MultiFormatWriter multiFormatWriter = new MultiFormatWriter();
                BitMatrix bitMatrix = null;

                if (barcodeFormat.Equals("CODE_128"))
                {
                    bitMatrix = multiFormatWriter.encode(texto, BarcodeFormat.CODE_128, height, width);
                }
                else if (barcodeFormat.Equals("EAN_8"))
                {
                    bitMatrix = multiFormatWriter.encode(texto, BarcodeFormat.EAN_8, height, width);
                }
                else if (barcodeFormat.Equals("EAN_13"))
                {
                    bitMatrix = multiFormatWriter.encode(texto, BarcodeFormat.EAN_13, height, width);
                }
                else if (barcodeFormat.Equals("PDF_417"))
                {
                    bitMatrix = multiFormatWriter.encode(texto, BarcodeFormat.PDF_417, height, width);
                }
                else if (barcodeFormat.Equals("QR_CODE"))
                {
                    bitMatrix = multiFormatWriter.encode(texto, BarcodeFormat.QR_CODE, height, width);
                }

                var bitWidth = bitMatrix.Width;
                var bitHeight = bitMatrix.Height;

                int[] pixelsImage = new int[bitWidth * bitHeight];

                //gerando o bitmap do barcodeformat com a largura, altura e formato de cor
                for (int i = 0; i < bitHeight; i++)
                {
                    for (int j = 0; j < bitWidth; j++)
                    {
                        if (bitMatrix[j, i])
                            pixelsImage[i * bitWidth + j] = (int)Convert.ToInt64(0xff000000);
                        else
                            pixelsImage[i * bitWidth + j] = (int)Convert.ToInt64(0xffffffff);
                    }
                }

                Bitmap bitmap = Bitmap.CreateBitmap(bitWidth, bitHeight, Bitmap.Config.Argb8888);
                bitmap.SetPixels(pixelsImage, 0, bitWidth, 0, 0, bitWidth, bitHeight);

                this.pictureConfig = new GEDI_PRNTR_st_PictureConfig();
                this.pictureConfig.Alignment = GEDI_PRNTR_e_Alignment.ValueOf(this.configPrint.Alinhamento);

                this.pictureConfig.Height = bitmap.Height;
                this.pictureConfig.Width = bitmap.Width;

                this.ImpressoraInit();
                this.iPrintr.DrawPictureExt(pictureConfig, bitmap);

                return true;
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }
        }

        public bool ImprimeImagem(string imagem)
        {
            int id;
            Bitmap bitmap;
            try
            {
                this.pictureConfig = new GEDI_PRNTR_st_PictureConfig();
                this.pictureConfig.Alignment = GEDI_PRNTR_e_Alignment.ValueOf(this.configPrint.Alinhamento);

                id = this.mainContext.Resources.GetIdentifier(imagem, "drawable", this.mainContext.PackageName);
                bitmap = BitmapFactory.DecodeResource(this.mainContext.Resources, id);
                if (modelo.Equals("Smart G800"))
                {
                    this.pictureConfig.Height = bitmap.Height;
                    this.pictureConfig.Width = bitmap.Width;
                }
                else
                {
                    this.pictureConfig.Height = configPrint.IHeight;
                    this.pictureConfig.Width = configPrint.IWidth;
                }
                ImpressoraInit();
                this.iPrintr.DrawPictureExt(pictureConfig, bitmap);

            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }

            return true;
        }

        public void ImprimeTexto(string texto)
        {
            try
            {
                ImpressoraInit();

                if (!this.IsImpressoraOK())
                {
                    throw new Exception(IMPRESSORA_ERRO);
                }
                sPrintLine(texto);
            }
            catch (Exception e)
            {
                throw new Exception(e.Message);
            }
        }

        public void ImprimeTexto(string texto, int tamanho)
        {
            int tamanhoOld;
            try
            {
                this.getStatusImpressora();
                if (!this.IsImpressoraOK())
                {
                    throw new Exception(IMPRESSORA_ERRO);
                }
                tamanhoOld = this.configPrint.Tamanho;
                this.configPrint.Tamanho = tamanho;
                sPrintLine(texto);
                this.configPrint.Tamanho = tamanhoOld;
            }
            catch (Exception e)
            {
                throw new Exception(e.Message);
            }
        }

        public void ImprimeTexto(string texto, bool negrito)
        {
            bool negritoOld;
            try
            {
                this.getStatusImpressora();
                if (!this.IsImpressoraOK())
                {
                    throw new Exception(IMPRESSORA_ERRO);
                }
                negritoOld = this.configPrint.Negrito;
                this.configPrint.Negrito = negrito;
                sPrintLine(texto);
                this.configPrint.Negrito = negritoOld;
            }
            catch (Exception e)
            {
                throw new Exception(e.Message);
            }
        }

        public void ImprimeTexto(string texto, bool negrito, bool italico)
        {
            bool negritoOld;
            bool italicoOld;
            try
            {
                this.getStatusImpressora();
                if (!this.IsImpressoraOK())
                {
                    throw new Exception(IMPRESSORA_ERRO);
                }
                negritoOld = this.configPrint.Negrito;
                italicoOld = this.configPrint.Italico;
                this.configPrint.Negrito = negrito;
                sPrintLine(texto);
                this.configPrint.Negrito = negritoOld;
                this.configPrint.Italico = italicoOld;
            }
            catch (Exception e)
            {
                throw new Exception(e.Message);
            }
        }

        public void ImprimeTexto(string texto, bool negrito, bool italico, bool sublinhado)
        {
            bool negritoOld;
            bool italicoOld;
            bool sublinhadoOld;
            try
            {
                this.getStatusImpressora();
                if (!this.IsImpressoraOK())
                {
                    throw new Exception(IMPRESSORA_ERRO);
                }
                negritoOld = this.configPrint.Negrito;
                italicoOld = this.configPrint.Italico;
                sublinhadoOld = this.configPrint.SubLinhado;
                this.configPrint.Negrito = negrito;
                sPrintLine(texto);
                this.configPrint.Negrito = negritoOld;
                this.configPrint.Italico = italicoOld;
                this.configPrint.SubLinhado = sublinhadoOld;
            }
            catch (Exception e)
            {
                throw new Exception(e.Message);
            }
        }

        public bool IsImpressoraOK()
        {
            if (this.status.Value == 0)
            {
                return true;
            }
            return false;
        }

        public void setConfigImpressao(ConfigPrint config)
        {
            this.configPrint = config;

            this.stringConfig = new GEDI_PRNTR_st_StringConfig(new Paint());
            this.stringConfig.Paint.TextSize = configPrint.Tamanho;
            this.stringConfig.Paint.TextAlign = Paint.Align.ValueOf(configPrint.Alinhamento);
            this.stringConfig.Offset = configPrint.OffSet;
            this.stringConfig.LineSpace = configPrint.LineSpace;

            switch (configPrint.Fonte)
            {
                case "NORMAL":
                    this.typeface = Typeface.Create(configPrint.Fonte, TypefaceStyle.Bold);
                    break;

                case "DEFAULT":
                    this.typeface = Typeface.Create(Typeface.Default, TypefaceStyle.Normal);
                    break;

                case "DEFAULT BOLD":
                    this.typeface = Typeface.Create(Typeface.DefaultBold, TypefaceStyle.Normal);
                    break;

                case "MONOSPACE":
                    this.typeface = Typeface.Create(Typeface.Monospace, TypefaceStyle.Normal);
                    break;

                case "SANS SERIF":
                    this.typeface = Typeface.Create(Typeface.SansSerif, TypefaceStyle.Normal);
                    break;

                case "SERIF":
                    this.typeface = Typeface.Create(Typeface.Serif, TypefaceStyle.Normal);
                    break;

                default:
                    this.typeface = Typeface.CreateFromAsset(this.mainContext.Assets, "fonts/" + configPrint.Fonte);
                    break;
            }

            if (this.configPrint.Negrito && this.configPrint.Italico)
            {
                this.typeface = Typeface.Create(typeface, TypefaceStyle.BoldItalic);
            }
            else if (this.configPrint.Negrito)
            {
                this.typeface = Typeface.Create(typeface, TypefaceStyle.Bold);
            }
            else if (this.configPrint.Italico)
            {
                this.typeface = Typeface.Create(typeface, TypefaceStyle.Italic);
            }

            if (this.configPrint.SubLinhado)
            {
                this.stringConfig.Paint.Flags = PaintFlags.UnderlineText;
            }

            this.stringConfig.Paint.SetTypeface(typeface);
        }

        public bool sPrintLine(string texto)
        {
            try
            {
                ImpressoraInit();
                this.iPrintr.DrawStringExt(this.stringConfig, texto);
                this.AvancaLinha(configPrint.AvancaLinha);
            }
            catch (GediException e)
            {
                throw new GediException(e.ErrorCode);
            }

            return true;
        }

        public string traduzStatusImpressora(GEDI_PRNTR_e_Status status)
        {

            string retorno;

            if (status == GEDI_PRNTR_e_Status.Ok)
            {
                retorno = "IMPRESSORA OK";
            }
            else if (status == GEDI_PRNTR_e_Status.OutOfPaper)
            {
                retorno = "SEM PAPEL";
            }
            else if (status == GEDI_PRNTR_e_Status.Overheat)
            {
                retorno = "SUPER AQUECIMENTO";
            }
            else
            {
                retorno = "ERRO DESCONHECIDO";
            }

            return retorno;

        }

    }
}